crypto/dh/dh_key.c

```diff
@@ -327,11 +327,11 @@ static int generate_key(DH *dh)
                 goto err;
 #else
             if (dh->params.q == NULL) {
-                /* secret exponent length, must satisfy 2^(l-1) <= p */
+                /* secret exponent length, must satisfy 2^l < (p-1)/2 */
                 if (dh->length != 0
-                    && dh->length >= BN_num_bits(dh->params.p))
+                    && dh->length >= BN_num_bits(dh->params.p) - 1)
```

Comment:
I'm wondering if this is a good idea or not. Maybe at some point we actually generated files that set it like that.